helper/schema/resource.go

```diff
@@ -195,21 +216,36 @@ func (r *Resource) ShimInstanceStateFromValue(state cty.Value) (*terraform.Insta
 	return s, nil
 }
 
-// See Resource documentation.
+// Deprecated: Please use the context aware equivalent CreateContextFunc.
 type CreateFunc func(*ResourceData, interface{}) error
 
-// See Resource documentation.
+// Deprecated: Please use the context aware equivalent ReadContextFunc.
 type ReadFunc func(*ResourceData, interface{}) error
 
-// See Resource documentation.
+// Deprecated: Please use the context aware equivalent UpdateContextFunc.
 type UpdateFunc func(*ResourceData, interface{}) error
 
-// See Resource documentation.
+// Deprecated: Please use the context aware equivalent DeleteContextFunc.
 type DeleteFunc func(*ResourceData, interface{}) error
 
-// See Resource documentation.
+// Deprecated: Please use the context aware equivalent ExistsContextFunc.
 type ExistsFunc func(*ResourceData, interface{}) (bool, error)
 
+// See Resource documentation.
+type CreateContextFunc func(context.Context, *ResourceData, interface{}) error
```

Comment:
If we are forcing a new function signature, can we please consider two things:

* The ability to passthrough Terraform warning and error diagnostics -- this might be implicitly handled with `error` (sorry cannot remember), but if not, this is a good opportunity to inject the concept into here (protocol 4 can just drop warnings and raise errors as `error` if some conversion is necessary)
* Having one of the input parameters types or the output result type (e.g. `CreateError` that thinly wraps `error`) differentiate between Create/Read/Update/Delete functions for static analysis -- since each of these types currently share the same underlying function signature, its not easily possible to determine which type of action function it is to write linting tooling.

Reply:
Hey Brian thanks for the feedback!
 * Yes there are plans to accomplish this, it would likely arrive in a follow up PR, but we can have developers returns a new error type (without changing the return type) and in `grpc_provider.go` try and cast to this enhanced type to give warnings/diags. We also discussed it being a pointer type (pointer, slice, map) attached to `ctx` so the developer can optionally fill it in and again we check it has been filled in `grpc_provider.go`.

* I'm not as up to speed on the challenges static analysis is facing. Even though all these funcs are types, I'm guessing because they are all the same signature your tooling is blocked. Without introducing something to the signature that differentiates them I'm not how this can be fixed? I see changing the signature in that way as less likely, as it might become hard to keep everything working with the shims. We will of course consider this though if the demand is there, it just might not arrive for a few iterations.